pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the :class:`~pennylane.data.DatasetBase` class, and `qml.data.Attribute` class
+for declaratively defining dataset classes.
+"""
+
+import typing
+from dataclasses import InitVar, dataclass
+from pathlib import Path
+from types import MappingProxyType
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Generic,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+    cast,
+    get_origin,
+)
+
+from typing_extensions import dataclass_transform
+
+from pennylane.data.base._zarr import zarr
+from pennylane.data.base.attribute import AttributeInfo, AttributeType
+from pennylane.data.base.mapper import MapperMixin, match_obj_type
+from pennylane.data.base.typing_util import UNSET, T, ZarrAny, ZarrGroup
+
+
+@dataclass
+class Attribute(Generic[T]):
+    """
+    The Attribute class is used to declaratively define the
+    attributes of a Dataset subclass, in a similar way to
+    dataclasses. This class should not be used directly when
+    declaring attributes, use the ``attribute()`` function instead.
+
+    Attributes:
+        attribute_type: The ``AttributeType`` class for this attribute
+        info: Attribute info
+        default: Default value for this attribute, if not provided
+            to __init__()
+        default_factory: Zero-argument callable that returns
+            a default value
+    """
+
+    attribute_type: Type[AttributeType[ZarrAny, T, Any]]
+    info: AttributeInfo
+    default: Union[Literal[UNSET], T, None]
+    default_factory: Optional[Callable[[], T]] = None
+
+
+def attribute(  # pylint: disable=too-many-arguments, unused-argument
```

Comment:
What does the workflow look like for creating a dataset with attributes that have documentation? Is it something like:
```
a = qml.data.attribute(attribute_type='int',doc='this is a test array', py_type='list')
dataset = qml.data.Dataset(v1 =a)
```
I couldn't figure this out.

Reply:
Reminder for when we work on documentation. It would be good to have an example of how to do this in the docstrings.